File: 0-frameworks-in-backend/1-java/src/main/java/com/example/dog/DogService.java
package com.example.dog;

import java.util.List;
import org.springframework.stereotype.Service;
import com.example.cat.Cat;
import com.example.cat.CatService;

// DogService depends on CatService (another package/"module") via constructor
// injection. No "new CatService()" — Spring's container injects it.)
@Service
public class DogService {
    private final CatService cats;

    public DogService(CatService cats) {
        this.cats = cats;
    }

    // records keep field order -> {mission, dependency, status} / {dog, borrowedCats}.
    public record SpyReport(String mission, String dependency, String status) {
    }

    public record CatsView(String dog, List<Cat> borrowedCats) {
    }

    public SpyReport spyReport() {
        return new SpyReport("cross-module-dependency-check", cats.spyHint(), "ok");
    }

    // reuses the same CatService bean (singleton) -> same /cats data.
    public CatsView catsViaDI() {
        return new CatsView("Rex", cats.all());
    }
}

File: 0-frameworks-in-backend/2-csharp/Dog/DogService.cs
using FrameworksDemo.Cat;

// the dog "module" — depends on CatService via constructor injection.
namespace FrameworksDemo.Dog;

public class DogService
{
    private readonly CatService _cats;

    // constructor injection: the container reads the parameter type and supplies
    // CatService. No "new CatService()" here (inversion of control).)
    public DogService(CatService cats) => _cats = cats;

    // the anonymous object keeps declaration order -> {mission, dependency, status}.
    public object SpyReport() => new
    {
        mission = "cross-module-dependency-check",
        dependency = _cats.SpyHint(),
        status = "ok",
    };

    // reuses the same CatService instance (registered Singleton) -> same /cats data.
    public object CatsViaDI() => new { dog = "Rex", borrowedCats = _cats.All() };
}
